gum/backend-mips/guminterceptor-mips.c

```diff
@@ -0,0 +1,550 @@
+  g_assert_cmpuint (gum_mips_writer_offset (cw),
+      <=, ctx->trampoline_slice->size);
+
+  ctx->overwritten_prologue_len = reloc_bytes;
+  memcpy (ctx->overwritten_prologue, function_address, reloc_bytes);
+
+  return TRUE;
+}
+
+void
+_gum_interceptor_backend_destroy_trampoline (GumInterceptorBackend * self,
+                                             GumFunctionContext * ctx)
+{
+  gum_code_slice_free (ctx->trampoline_slice);
+  gum_code_deflector_free (ctx->trampoline_deflector);
+  ctx->trampoline_slice = NULL;
+  ctx->trampoline_deflector = NULL;
+}
+
+void
+_gum_interceptor_backend_activate_trampoline (GumInterceptorBackend * self,
+                                              GumFunctionContext * ctx,
+                                              gpointer prologue)
+{
+  GumMipsWriter * cw = &self->writer;
+  GumMipsFunctionContextData * data = (GumMipsFunctionContextData *)
+      &ctx->backend_data;
+  GumAddress on_enter = GUM_ADDRESS (ctx->on_enter_trampoline);
+
+  gum_mips_writer_reset (cw, prologue);
+  cw->pc = GUM_ADDRESS (ctx->function_address);
+
+  if (ctx->trampoline_deflector != NULL)
+  {
+    /* TODO: implement branch to deflector */
+    g_assert_not_reached ();
+  }
+  else
+  {
+    switch (data->redirect_code_size)
+    {
+      case 8:
+        gum_mips_writer_put_j_address (cw, on_enter);
+        break;
+      case 16:
+        gum_mips_writer_put_la_reg_address (cw, MIPS_REG_AT, on_enter);
+        gum_mips_writer_put_jr_reg (cw, MIPS_REG_AT);
+        break;
+      default:
+        g_assert_not_reached ();
+    }
+  }
+
+  gum_mips_writer_flush (cw);
+  g_assert_cmpuint (gum_mips_writer_offset (cw), <=, data->redirect_code_size);
+}
+
+void
+_gum_interceptor_backend_deactivate_trampoline (GumInterceptorBackend * self,
+                                                GumFunctionContext * ctx,
+                                                gpointer prologue)
+{
+  (void) self;
+
+  memcpy (prologue, ctx->overwritten_prologue, ctx->overwritten_prologue_len);
+}
+
+gpointer
+_gum_interceptor_backend_get_function_address (GumFunctionContext * ctx)
+{
+  return ctx->function_address;
+}
+
+gpointer
+_gum_interceptor_backend_resolve_redirect (GumInterceptorBackend * self,
+                                           gpointer address)
+{
+  /* TODO: implement resolve redirect */
+  return NULL;
+}
+
+gpointer
+_gum_interceptor_invocation_get_nth_argument (GumInvocationContext * context,
+                                              guint n)
+{
+  if (n < 4)
+  {
+    switch (n)
+    {
+      case 0:
+        return (gpointer) context->cpu_context->a0;
+      case 1:
+        return (gpointer) context->cpu_context->a1;
+      case 2:
+        return (gpointer) context->cpu_context->a2;
+      case 3:
+        return (gpointer) context->cpu_context->a3;
+    }
+  }
+  else
+  {
+    gpointer * stack_argument = (gpointer *) context->cpu_context->sp;
+
+    return stack_argument[n - 4];
+  }
+
+  return NULL;
```

Comment:
It's unreachable, but if you don't put a return here, you get a compiler warning... I prefer to get rid of the compiler warning. What do you think?